tests/CacheCompat/CommonCache.Test.Unit/CacheExecutionTests.cs

```diff
@@ -27,96 +27,132 @@
 
-        [TestMethod]
-        public async Task AdalV4ToAdalV3Async()
+        [DataTestMethod]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.AdalV3, CacheStorageType.Adal,   DisplayName = "AdalV4->AdalV3 adal v3 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.AdalV4, CacheStorageType.Adal,   DisplayName = "AdalV4->AdalV4 adal v3 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.AdalV5, CacheStorageType.Adal,   DisplayName = "AdalV4->AdalV5 adal v3 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.AdalV4, CacheStorageType.MsalV2, DisplayName = "AdalV4->AdalV4 msal v2 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.AdalV5, CacheStorageType.MsalV2, DisplayName = "AdalV4->AdalV5 msal v2 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.MsalV2, CacheStorageType.Adal,   DisplayName = "AdalV4->MsalV2 adal v3 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.MsalV2, CacheStorageType.MsalV2, DisplayName = "AdalV4->MsalV2 msal v2 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.MsalV3, CacheStorageType.Adal,   DisplayName = "AdalV4->MsalV3 adal v3 cache")]
+        [DataRow(CacheProgramType.AdalV4, CacheProgramType.MsalV3, CacheStorageType.MsalV2, DisplayName = "AdalV4->MsalV3 msal v2 cache")]
+        public async Task TestAdalV4CacheCompatibilityAsync(
+            CacheProgramType interactiveType,
+            CacheProgramType silentType,
+            CacheStorageType cacheStorageType)
         {
-            // Sign in via adal v4, token is written in cache with old adal v3 cache format (expected size 5156 bytes)
-            // Run adal v3, token comes from cache as adalv3 should understand cache format.
             var executor = new CacheTestExecutor(
-                CacheProgramType.AdalV4,
-                CacheProgramType.AdalV3,
-                5156,
+                interactiveType,
+                silentType,
+                cacheStorageType,
                 expectSecondTokenFromCache: true);
 
             await executor.ExecuteAsync(CancellationToken.None).ConfigureAwait(false);
         }
 
-        [TestMethod]
-        public async Task AdalV4ToMsalV2Async()
+        [DataTestMethod]
+        //[DataRow(CacheProgramType.AdalV5, CacheProgramType.AdalV3, CacheStorageType.Adal,   DisplayName = "AdalV5->AdalV3 adal v3 cache")]
+        //[DataRow(CacheProgramType.AdalV5, CacheProgramType.AdalV4, CacheStorageType.Adal,   DisplayName = "AdalV5->AdalV4 adal v3 cache")]
+        //[DataRow(CacheProgramType.AdalV5, CacheProgramType.AdalV5, CacheStorageType.Adal,   DisplayName = "AdalV5->AdalV5 adal v3 cache")]
+        //[DataRow(CacheProgramType.AdalV5, CacheProgramType.AdalV4, CacheStorageType.MsalV2, DisplayName = "AdalV5->AdalV4 msal v2 cache")]
+        [DataRow(CacheProgramType.AdalV5, CacheProgramType.AdalV5, CacheStorageType.MsalV2, DisplayName = "AdalV5->AdalV5 msal v2 cache")]
```

Comment:
I think this is fine as long as we don't expect these tests to fail too often. The downside of data driven tests is that when one of the row fails, it's not immediately apparent which row actually caused the failure, you have to look into the details. You also have to write a lot of DataRow lines

We have the same problem with UI tests, where we have 2 dimentions we want to test: user (AAD, ADFSv2, v3, v4, B2C), cloud (normal, DE, US, CN), and will probably grow in time. I was thinking of using T4 to generate the tests (i.e. the method signatures and the paarams) instead of DataRow. 

@trwalke FYI

Reply:
The "message" on the end is actually very helpful in my debugging as you can see them specified there in the test output.  I also broke this down into groups (adalv3, adalv4, adalv5, etc) so that we have less tests to run in a particular block.  It's a good concern, but with this many permutations it was gfetting very difficult to ensure and easily see that all of the test cases were properly covered without this approach.